README

	                                ____  ____  ____  ____  ____
	Scheme 9 from Empty Space      / ___)/ __ \|  __)|  __)/ ___)
	A Portable Scheme Interpreter  \___ \\__  /|  __)|  __)\___ \
	By Nils M Holm, 2007-2018      (____/(___/ |_|   |____)(____/

	S9fES is a mature, portable, and comprehensible public-domain
	interpreter for R4RS Scheme offering

	- bignum arithmetics
	- decimal-based real number arithmetics
	- support for low-level Unix programming
	- cursor addressing with Curses
	- basic networking procedures
	- an integrated online help system
	- loads of useful library functions

	It is written in ANSI C (C89) and Scheme and runs in many
	popular environments, including *BSD, Linux, Plan 9, and the
	unmentionable horror (although the Plan 9 and Horror versions
	will exclude the Unix, Curses, and Networking extensions).

	The S9fES code strives to be simple and comprehensible. It is
	particularly interesting to people who want to

	- have a Scheme system that runs virtually everywhere
	- write Unix programs in a high-level language
	- try Scheme without having to jump through too many hoops
	- study the implementation of Scheme

	There is a book describing a previous S9 implementation
	in detail. It is available in print and PDF format. See
	http://t3x.org/s9book/

	S9fES supports the following SRFI's: 
	- SRFI-0:  feature-based conditional expansion (subset)
	- SRFI-2:  AND-LET*                            (subset)
	- SRFI-22: running Scheme scripts on Unix
	- SRFI-23: error reporting mechanism
	- SRFI-30: nested multi-line comments
	- SRFI-43: vector library                      (subset)

	Bakul Shah maintains a downstream repository with extended
	Plan 9 support, like access to system functions and an FFI:
	https://github.com/bakul/s9fes.  Please send any feedback
	regarding the Plan 9 extension directly to him.


	QUICK START

	You can run an S9 interpreter in its build directory without
	installing it. Just type

	    cc -o s9 s9.c s9core.c
	    ./s9 -i - -d s9.image

	and then "./s9". You might have to add -Dplan9 on Plan 9 systems.
	The interpreter will support full R4RS Scheme, but none of the
	extensions provided by S9 (Unix system calls, Curses, etc).

	A quick test of the interpreter can be performed by running

	    ./s9 util/test.scm

	In order to bring up the online help system, run

	    ./s9 -l contrib/help.scm -d s9.image

	When running ./s9 for the next time, the following help commands
	will be available:

	    ,h  explain the online help system
	    ,a  return a list of all help topics (long!)

	On most systems of the Unix family (plus CygWin and MinGW),
	you can compile and install S9fES by running "make install".

	See below for further instructions.


	COMPILING AND INSTALLING

	Unix

	To compile S9fES, run "make" and then "make test" (this will run
	the test suite to make sure that the interpreter works properly).
	Running "make tests" will run even more tests.

	You can install S9fES (including the goodies) on a Unix system
	by typing "make install" as root, but before doing so, please
	edit at least the PREFIX variable at the beginning of Makefile.
	Be sure to re-compile S9fES (make clean; make) after changing
	PREFIX, because it is used to set up some internal paths.

	Plan 9

	To compile S9fES on Plan 9 from Bell Labs, just type "mk".
	The "test" and "tests" targets will run tests, just like on
	Unix.

	To install S9fES, run "mk inst". It will copy the S9fES binary,
	heap image, function library, and help files to $s9dir (which
	defaults to /lib/s9fes). Edit mkfile if you prefer a different
	location.

	The install procedure will also copy the file util/s9.rc to
	/rc/bin/s9 and adjust the $s9dir variable in that file.

	Manual Installation

	To install S9fES manually,

	- Compile s9 with a proper image directory and library path (the
	  ones hardwired in "s9.c" are probably not what you want). E.g.:

	  cc -o s9 -DIMAGE_DIR="\"dir\"" -DLIBRARY_PATH="\"path\"" \
		s9.c s9core.c

	  where "dir" should be set to S9DIR, as described below, and a
	  reasonable value for "path" would be

	  ".:~/s9fes:/usr/local/share/s9fes"

	  Security-sensitive people may consider (re)moving the dot.

	- Copy the "s9" binary to a location where it can be executed
	  (e.g. /usr/local/bin).

	- Select a publicly readable directory to contain library files.
	  (default: /usr/local/share/s9fes). This directory will be
	  referred to as S9DIR in the following instructions. Note
	  that this directory *must* be contained in LIBRARY_PATH,
	  as defined above.

	- Compile a heap image and copy it to S9DIR:

	  s9 -i - -d s9.image

	  This step requires the file "s9.scm" to be present in the
	  current directory.

	  The image file must be named "s9.image" or it will not be
	  loaded by default. An image loads significantly faster than
	  source code.

	Optionally:

	- Copy the content of the "lib" directory to S9DIR. This
	  directory contains lots of useful Scheme functions.

	- Copy the content of the "contrib" directory to S9DIR. These
	  files contain additional Scheme functions contributed by other
	  authors and/or imported from various sources.

	- Create a subdirectory named "help" in S9DIR and copy the
	  content of the "help" directory to S9DIR/help. These files
	  are part of the interactive help system. Note that help/
	  does contain subdirectories!

	- Copy the nroff(1) source code of the manual page "s9.1" to
	  section 1 of your online manual (e.g. /usr/local/man/man1).
	  In case you are not running Unix, there is a pre-formatted
	  copy in the file "s9.1.txt".


	CONFIGURATION

	To create an image file containing additional functionality,
	add the desired options to the "config.scm" file, run

	s9 -i - -l config.scm -d s9.image

	and copy the resulting image to your IMAGE_DIR.

	If you are keeping your image files in a location other than
	S9DIR, you can create the S9FES_IMAGE_DIR environment variable
	and make it point to your preferred location. In this case
	"s9 -i foo.image" will load $S9FES_IMAGE_DIR/foo.image.

	You may also create the S9FES_LIBRARY_PATH environment variable
	and make it point to S9DIR as well as other directories that
	contain Scheme programs. The variable may contain a single
	directory or a colon-separated list of directories that will be
	searched in sequence for library files and help pages (in "help"
	subdirectories contained in the given paths).

	S9FES_LIBRARY_PATH overrides LIBRARY_PATH, so all directories
	listed in the latter should also be present in the former.


	GETTING STARTED

	Typing "s9" will drop you into the read-eval-print loop of the
	interpreter. You can run Scheme programs non-interactively by
	typing "s9 program.scm" at your shell prompt.

	If you installed the extension library functions in S9DIR,
	they can be loaded by the LOAD-FROM-LIBRARY procedure or the
	",l" (comma ell) meta command:

	> ,l draw-tree
	; loading from /usr/local/share/s9fes/draw-tree.scm
	> (draw-tree '(a b c))
	[o|o]---[o|o]---[o|/]
	 |       |       |
	 a       b       c
	> _

	Feel free to explore them.

	Running "s9help topic" on the shell prompt will print the S9fES
	online help page about the given topic.


	ACKNOWLEDGEMENTS

	I would like to thank the following people and organisations for
	patches, bug reports, suggestions, hardware, access to hardware,
	etc:

	Alexander Shendi, Arun Srinivasan, Bakul Shah,
	Barak Pearlmutter, Blake McBride, Bsamograd (reddit),
	David Person, Dig1 (reddit), Dirk Lutzebaeck,
	Doug Currie, Mario Deilmann, Masaru KIMURA,
	McKay Marston, Nullbuilt0 (reddit), Peter, Ray Lai,
	Torsten Leibold, Victor M. Muriel, Yi Dai,
	and the Super Dimension Fortress (SDF.ORG).


	CONTACT

	Nils M Holm < n m h @ t 3 x . o r g >



HISTORY

2019-04-02

	Imported latest S9core. No user-visible changes.

2018-12-05

	Merged S9core fix by Micah Stetsion: s9core GC should never close
	the current input or output port. Does not affect Scheme 9, but
	is might affect other languages (like Klong). Thanks, Micah!

	Fix: APPLYing a function to more than CHUNK_SIZE (1024) arguments
	would result in a core dump. (Not that doing this is a good idea,
	but it should work none the less).  The fix is SLOW. Be warned!

2018-11-24

	Fix: DISPLAY, WRITE, and REPL output could not be interrupted
	by pressing INTR. (s9.c)

2018-11-19

	Exported tag() macro of S9core. (s9import.h)

	Removed unused "Error" variable and misc. cosmetic changes.
	(s9core.h)

2018-11-15

	Removed 16-bit support. It was mostly a lie anyway. (s9core.h)

	Applied various fixes to TROFF mode in EDOC. (prog/)

2018-11-14

	Fix: EDOC generated index commands that were not contained in
	separate lines in TROFF mode. (prog/)

	Fix: C2HTML created spurious \fP in TROFF mode. (contrib/)

	Fix: ptrdiff_t should be vlong on 64-bit Plan 9. (s9core.h)

2018-11-13

	Fix: LOCATE-FILE did not decode leading ~ in path names properly.
	(s9.scm)

	Misc. code clean-up. (s9.c, s9core.*)

2018-11-12

	Added S9core self-test to Plan 9 tests. (mkfile)

	Fixed some minor bugs that kept S9 from compiling on Plan 9.
	(s9core.h)

2018-11-11

	Updated TROFF macro package for EDOC. (prog/)

	Corrected LOCATE-FILE help page. (help/)

	Fix: REVERSE! can no longer reverse immutable lists. (s9.c)

	Removed left-over help page. (help/)

	Added S9core self-test to tests. (Makefile)

	S9core: adjusted size of internal GC protection stack, because
	real_power() can recurse frequently. (Note: real_power() is not
	used by S9fES). (s9core.c)

2018-11-09

	Merged Linux/POSIX patches from Barak Pearlmutter's
	downstream repository (http://github.com/barak/scheme9).
	Thanks!

	Disabled some system extension tests that failed on
	some platforms, probably due to wrong use of SYS:FORK.
	To be investigated! (util/systest.scm)

2018-11-08

	Added TROFF output mode to EDOC. (prog/)

2018-11-06

	Added TROFFIFY-CHAR function for escaping dangerous chars in
	TROFF. (lib/)

2018-11-05

	EVAL context is now saved per file in LOAD, instead of per
	expression. (s9.c)

2018-11-04

	S9core uses a more efficient way to GC-protect internal nodes
	now, resulting in a 20% to 30% increase in real and bignum
	performance. (s9core.c)

	STRING-APPEND and VECTOR-APPEND now allocate only one single
	fresh vector when appending multiple arguments. They basically
	folded over their arguments before. (s9.c)

2018-11-03

	The / procedure signals an error now when the divisor is 0.
	Use CATCH-ERRORS to intercept this condition and deliver
	some value. (s9.c)

	(Expt 0 0.0) and (log x) for non-positive x now signal a range
	error. (s9.scm)

	Fixed some type mismatches in the 64-bit build.

	Removed 64-bit build option from Makefile and README.
	*** Note: S9 will still build on 64-bit systems, but will use
	*** 32-bit structures internally. This should not make any
	*** difference to anybody. If it does, let me know!

2018-11-02

	Factored out MAKE-HELP-INDEX procedure. (lib/)

	LIST->SET now uses a hash table when its argument exceeds
	a certain size.

	Misc. small changes to S9core. (s9core.*, s9import.h)

2018-11-01

	Cleaned up help directory: removed left-over pages, linked
	missing pages. Reorganized the help system structure so that
	no links, symlinks, or duplicate files are needed. INDEX
	files are now local to subdirectories of help/.

	Retired FIND-HELP-PATH and FIND-HELP procedures from the
	SYS-UNIX extension.

	Retired S9SYMBOLS command. (prog/)

	Added FIND-HELP-PAGE procedure, which returns the full path
	of the help file describing a given syntactic keyword or
	procedure (or #F if no such page exists). (lib/)

	Added SCAN-HELP-PAGES procedure, which searches all help pages
	for a given keyword. This procedure replaces FIND-HELP. Unlike
	FIND-HELP it does not require the SYS-UNIX extension. (lib/)

	Updated S9HELP command and added -c (check database) option.
	(prog/)

-------------------------------------------------------------------------------
	Major Change: argument passing to extension procedures
-------------------------------------------------------------------------------

2018-10-31

	As of this release, arguments to extension procedures (extprocs)
	are passed on the stack. In functions implementing extprocs, the
	n'th argument is now retrieved by calling parg(n). The number
	of arguments is returned by narg(). The extproc handler type
	was changed from  cell (*f)(cell x)  to  cell (*f)(void).

	All extproc handlers need to be adapted to use the new calling
	conventions. The csv, curses, and sys-unix extensions already
	have been modified accordingly.

	Rationale:

	In previous versions of S9fES arguments were passed as lists.
	This method was still used in S9 Reimagined when applying
	extension procecures. However, this approach required to
	convert arguments on the stack to lists before passing them
	to extension procedures. This caused unnecessary consing.

2018-10-30

	Retired Plan 9 system extension (ext/sys-plan9). Please see
	https://github.com/bakul/s9fes for a more up to date version.

	Note that S9fES will continue to compile and run on Plan 9,
	but the stock version will no longer provide functions like
	SYS:OPEN, SYS:BIND, etc. Use Bakul's version if you need
	those!

	Replaced symlinks in help/ with links (which will hopefully
	extract to copies on non-Unix systems).

2018-10-29

	Added bytecode disassembler. (lib/disassemble.scm)

	Fixed a bug in the library test suite ("make libtest") that
	would report a failed image load when upgrading to a new
	version. (util/)

2018-10-28

	Added system() function for Plan 9 to S9core. Thanks, Bakul!

	Fix: PROCEDURE? returned #F for extension procedures. (s9.c)

	Retired lib/catch.scm (moved from lib/ to attic/).

2018-10-27


s9core.h
/*
 * S9core Toolkit, Mk IVd
 * By Nils M Holm, 2007-2019
 * In the public domain
 *
 * Under jurisdictions without a public domain, the CC0 applies:
 * https://creativecommons.org/publicdomain/zero/1.0/
 */

#define S9_VERSION "20190402"

/*
 * Ugly prelude to deal with some system-dependent stuff.
 */

#ifdef __NetBSD__
 #ifndef unix
  #define unix
 #endif
#endif

#ifdef __unix
 #ifndef unix
  #define unix
 #endif
#endif

#ifdef __linux
 #ifndef unix
  #define unix
 #endif
#endif

#ifdef __GNUC__
 #ifndef unix
  #define unix
 #endif
#endif

#ifdef __clang__
 #ifndef unix
  #define unix
 #endif
#endif

#ifndef unix
 #ifndef plan9
  #error "Either 'unix' or 'plan9' must be #defined."
 #endif
#endif

#ifdef _MSC_VER
 #if _MSC_VER > 1200
  #ifndef _CRT_SECURE_NO_DEPRECATE
   #define _CRT_SECURE_NO_DEPRECATE
  #endif
 #endif
 #ifndef _POSIX_
  #define _POSIX_
 #endif
#endif

#ifdef plan9
 #include <u.h>
 #include <libc.h>
 #include <stdio.h>
 #include <ctype.h>
 #include <ape/limits.h>
 #define size_t	uvlong
 #define bye(x)	exits((x)? "error": NULL)
 #define ptrdiff_t vlong
#endif

#ifdef unix
 #include <stdlib.h>
 #include <stddef.h>
 #include <stdio.h>
 #include <string.h>
 #include <ctype.h>
 #include <limits.h>
 #define bye(x)	exit((x)? EXIT_FAILURE: EXIT_SUCCESS)
#endif

/*
 * Tunable parameters
 */

/* Default memory limits in K nodes and K cells; 0 = none */
#define S9_NODE_LIMIT	14013
#define S9_VECTOR_LIMIT	14013

/* Initial memory pool sizes */
#define S9_INITIAL_SEGMENT_SIZE	32768

/* Primitive segment size (slots) */
#define S9_PRIM_SEG_SIZE	256

/* Maximum number of open I/O ports */
#define S9_MAX_PORTS		32

/* Use 64-bit cells (don't do this!) */
/* #define S9_BITS_PER_WORD_64 */

/*
 * Non-tunable parameters
 */

/* A cell must be large enough to hold an integer segment;
 * see S9_INT_SEG_LIMIT, below
 */

#ifdef S9_BITS_PER_WORD_64
 #define s9_cell	ptrdiff_t
#else
 #define s9_cell	int
#endif

#ifdef S9_BITS_PER_WORD_64
 #define S9_DIGITS_PER_CELL	18
 #define S9_INT_SEG_LIMIT	1000000000000000000LL
 #define S9_MANTISSA_SEGMENTS	1
#else
 #define S9_DIGITS_PER_CELL	9
 #define S9_INT_SEG_LIMIT	1000000000L
 #define S9_MANTISSA_SEGMENTS	2
#endif

#define S9_MANTISSA_SIZE	(S9_MANTISSA_SEGMENTS * S9_DIGITS_PER_CELL)

/*
 * Node tags
 */

#define S9_ATOM_TAG	0x01	/* Atom, car = type, cdr = next */
#define S9_MARK_TAG	0x02	/* Mark */
#define S9_STATE_TAG	0x04	/* State */
#define S9_VECTOR_TAG	0x08	/* Vector, car = type, cdr = content */
#define S9_PORT_TAG	0x10	/* Atom is an I/O port (with ATOM_TAG) */
#define S9_USED_TAG	0x20	/* Port: in use */
#define S9_LOCK_TAG	0x40	/* Port: locked (do not finalize) */
#define S9_CONST_TAG	0x80	/* Node is immutable */

/*
 * Special objects
 */

#define s9_special_p(x)	((x) < 0)
#define S9_NIL		(-1)
#define S9_TRUE		(-2)
#define S9_FALSE	(-3)
#define S9_END_OF_FILE	(-4)
#define S9_UNDEFINED	(-5)
#define S9_UNSPECIFIC	(-6)
#define S9_VOID		(-7)

/*
 * Type tags
 */

#define S9_T_ANY		(-10)
#define S9_T_BOOLEAN		(-11)
#define S9_T_CHAR		(-12)
#define S9_T_INPUT_PORT		(-13)
#define S9_T_INTEGER		(-14)
#define S9_T_LIST		(-17)
#define S9_T_OUTPUT_PORT	(-15)
#define S9_T_PAIR		(-16)
#define S9_T_PRIMITIVE		(-18)
#define S9_T_FUNCTION		(-19)
#define S9_T_REAL		(-20)
#define S9_T_STRING		(-21)
#define S9_T_SYMBOL		(-22)
#define S9_T_SYNTAX		(-23)
#define S9_T_VECTOR		(-24)
#define S9_T_CONTINUATION	(-25)
#define S9_T_FIXNUM		(-26)
#define S9_T_NONE		(-99)

#define S9_USER_SPECIALS	(-100)

/*
 * Structures
 */

struct S9_counter {
	int	n, n1k, n1m, n1g, n1t;
};

#define s9_counter	struct S9_counter

struct S9_primitive {
	char	*name;
	s9_cell	(*handler)(void);
	int	min_args;
	int	max_args;	/* -1 = variadic */
	int	arg_types[3];
};

#define S9_PRIM    struct S9_primitive

/*
 * I/O
 */

#define s9_nl()		s9_prints("\n")

/*
 * Nested lists
 */

#define s9_car(x)          (S9_car[x])
#define s9_cdr(x)          (S9_cdr[x])
#define s9_caar(x)         (S9_car[S9_car[x]])
#define s9_cadr(x)         (S9_car[S9_cdr[x]])
#define s9_cdar(x)         (S9_cdr[S9_car[x]])
#define s9_cddr(x)         (S9_cdr[S9_cdr[x]])
#define s9_caaar(x)        (S9_car[S9_car[S9_car[x]]])
#define s9_caadr(x)        (S9_car[S9_car[S9_cdr[x]]])
#define s9_cadar(x)        (S9_car[S9_cdr[S9_car[x]]])
#define s9_caddr(x)        (S9_car[S9_cdr[S9_cdr[x]]])
#define s9_cdaar(x)        (S9_cdr[S9_car[S9_car[x]]])
#define s9_cdadr(x)        (S9_cdr[S9_car[S9_cdr[x]]])
#define s9_cddar(x)        (S9_cdr[S9_cdr[S9_car[x]]])
#define s9_cdddr(x)        (S9_cdr[S9_cdr[S9_cdr[x]]])
#define s9_caaaar(x)       (S9_car[S9_car[S9_car[S9_car[x]]]])
#define s9_caaadr(x)       (S9_car[S9_car[S9_car[S9_cdr[x]]]])
#define s9_caadar(x)       (S9_car[S9_car[S9_cdr[S9_car[x]]]])
#define s9_caaddr(x)       (S9_car[S9_car[S9_cdr[S9_cdr[x]]]])
#define s9_cadaar(x)       (S9_car[S9_cdr[S9_car[S9_car[x]]]])
#define s9_cadadr(x)       (S9_car[S9_cdr[S9_car[S9_cdr[x]]]])
#define s9_caddar(x)       (S9_car[S9_cdr[S9_cdr[S9_car[x]]]])
#define s9_cadddr(x)       (S9_car[S9_cdr[S9_cdr[S9_cdr[x]]]])
#define s9_cdaaar(x)       (S9_cdr[S9_car[S9_car[S9_car[x]]]])
#define s9_cdaadr(x)       (S9_cdr[S9_car[S9_car[S9_cdr[x]]]])
#define s9_cdadar(x)       (S9_cdr[S9_car[S9_cdr[S9_car[x]]]])
#define s9_cdaddr(x)       (S9_cdr[S9_car[S9_cdr[S9_cdr[x]]]])
#define s9_cddaar(x)       (S9_cdr[S9_cdr[S9_car[S9_car[x]]]])
#define s9_cddadr(x)       (S9_cdr[S9_cdr[S9_car[S9_cdr[x]]]])
#define s9_cdddar(x)       (S9_cdr[S9_cdr[S9_cdr[S9_car[x]]]])
#define s9_cddddr(x)       (S9_cdr[S9_cdr[S9_cdr[S9_cdr[x]]]])

/*
 * Access to fields of atoms
 */

#define s9_tag(n)		(S9_tag[n])

#define s9_string(n)		((char *) &Vectors[S9_cdr[n]])
#define s9_string_len(n)	(Vectors[S9_cdr[n] - 1])
#define s9_symbol_name(n)	(string(n))
#define s9_symbol_len(n)	(string_len(n))
#define s9_vector(n)		(&Vectors[S9_cdr[n]])
#define s9_vector_link(n)	(Vectors[S9_cdr[n] - 3])
#define s9_vector_index(n)	(Vectors[S9_cdr[n] - 2])
#define s9_vector_size(k)	(((k)+sizeof(s9_cell)-1) / \
                                 sizeof(s9_cell) + 3)
#define s9_vector_len(n)	(vector_size(string_len(n)) - 3)

#define s9_fixval(x)		cadr(x)
#define s9_small_int_value(x)	cadr(x)
#define s9_port_no(n)		(cadr(n))
#define s9_char_value(n)	(cadr(n))
#define s9_prim_slot(n)		(cadr(n))
#define s9_prim_info(n)		(&Primitives[prim_slot(n)])

/*
 * Type predicates
 */

#define s9_eof_p(n)		((n) == S9_END_OF_FILE)
#define s9_undefined_p(n)	((n) == S9_UNDEFINED)
#define s9_unspecific_p(n)	((n) == S9_UNSPECIFIC)

#define s9_boolean_p(n)	\
	((n) == S9_TRUE || (n) == S9_FALSE)

#define s9_integer_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_ATOM_TAG) && car(n) == S9_T_INTEGER)

#define s9_real_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_ATOM_TAG) && car(n) == S9_T_REAL)

#define s9_fix_p(n) \
        (!s9_special_p(n) && (tag(n) & S9_ATOM_TAG) && T_FIXNUM == car(n))

#define s9_number_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_ATOM_TAG) && \
	 (car(n) == S9_T_REAL || car(n) == S9_T_INTEGER))

#define s9_primitive_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_ATOM_TAG) && \
	 car(n) == S9_T_PRIMITIVE)

#define s9_function_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_ATOM_TAG) && \
	 car(n) == S9_T_FUNCTION)

#define s9_continuation_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_ATOM_TAG) && \
	 car(n) == S9_T_CONTINUATION)

#define s9_char_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_ATOM_TAG) && car(n) == S9_T_CHAR)

#define s9_syntax_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_ATOM_TAG) && car(n) == S9_T_SYNTAX)

#define s9_input_port_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_ATOM_TAG) && \
	 (tag(n) & S9_PORT_TAG) && car(n) == S9_T_INPUT_PORT)

#define s9_output_port_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_ATOM_TAG) && \
	 (tag(n) & S9_PORT_TAG) && car(n) == S9_T_OUTPUT_PORT)

#define s9_symbol_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_VECTOR_TAG) && car(n) == S9_T_SYMBOL)

#define s9_vector_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_VECTOR_TAG) && car(n) == S9_T_VECTOR)

#define s9_string_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_VECTOR_TAG) && car(n) == S9_T_STRING)

#define s9_constant_p(n) \
	(!s9_special_p(n) && (tag(n) & S9_CONST_TAG))

#define s9_atom_p(n) \
	(s9_special_p(n) || (tag(n) & S9_ATOM_TAG) || (tag(n) & S9_VECTOR_TAG))

#define s9_pair_p(x) (!s9_atom_p(x))

#define s9_small_int_p(n) (NIL == cddr(n))

#define s9_type_tag(n) \
	(S9_TRUE == (n)? S9_T_BOOLEAN: \
	 S9_FALSE == (n)? S9_T_BOOLEAN: \
	 (!s9_special_p(n) && (tag(n) & (S9_ATOM_TAG|S9_VECTOR_TAG))? car(n): \
	 S9_T_NONE))

/*
 * Allocators
 */

#define s9_cons(pa, pd)		s9_cons3((pa), (pd), 0)
#define s9_new_atom(pa, pd)	s9_cons3((pa), (pd), S9_ATOM_TAG)
#define s9_save(n)		(Stack = s9_cons((n), Stack))

/*
 * Bignum arithmetics
 */

#define s9_bignum_negative_p(a)	((cadr(a)) < 0)
#define s9_bignum_zero_p(a)	((cadr(a)) == 0)
#define s9_bignum_positive_p(a)	((cadr(a)) > 0)

/*
 * Real number structure
 */

#define S9_real_flags(x)	(cadr(x))
#define S9_real_exponent(x)	(caddr(x))
#define S9_real_mantissa(x)	(cdddr(x))

#define S9_REAL_NEGATIVE   0x01

#define S9_real_negative_flag(x)	(S9_real_flags(x) & S9_REAL_NEGATIVE)

/*
 * Real-number arithmetics
 */

#define S9_real_zero_p(x) \
	(car(S9_real_mantissa(x)) == 0 && cdr(S9_real_mantissa(x)) == S9_NIL)

#define S9_real_negative_p(x) \
	(S9_real_negative_flag(x) && !S9_real_zero_p(x))

#define S9_real_positive_p(x) \
	(!S9_real_negative_flag(x) && !S9_real_zero_p(x))

#define S9_real_negate(a) \
	S9_make_quick_real(S9_real_flags(a) & S9_REAL_NEGATIVE?	\
			S9_real_flags(a) & ~S9_REAL_NEGATIVE: \
			S9_real_flags(a) |  S9_REAL_NEGATIVE, \
			S9_real_exponent(a), S9_real_mantissa(a))

/*
 * Globals
 */

extern s9_cell	*S9_car,
		*S9_cdr;
extern char	*S9_tag;

extern s9_cell	*S9_vectors;

extern s9_cell	S9_stack;

extern s9_cell	*S9_gc_stack;
extern int	*S9_gc_stkptr;

extern S9_PRIM	*S9_primitives;

extern s9_cell	S9_zero,
		S9_one,
		S9_two,
		S9_ten;

extern s9_cell	S9_epsilon;

extern FILE	*S9_ports[];

extern int	S9_input_port,
		S9_output_port,
		S9_error_port;

/*
 * Prototypes
 */

#ifdef plan9
int	system(char *s);
#endif

void	s9_abort(void);
int	s9_aborted(void);
void	s9_add_image_vars(s9_cell **v);
s9_cell	s9_apply_prim(s9_cell f);
s9_cell	s9_argv_to_list(char **argv);
long	s9_asctol(char *s);
s9_cell	s9_bignum_abs(s9_cell a);
s9_cell	s9_bignum_add(s9_cell a, s9_cell b);
s9_cell	s9_bignum_divide(s9_cell a, s9_cell b);
int	s9_bignum_equal_p(s9_cell a, s9_cell b);
int	s9_bignum_even_p(s9_cell a);
int	s9_bignum_less_p(s9_cell a, s9_cell b);
s9_cell	s9_bignum_multiply(s9_cell a, s9_cell b);
s9_cell	s9_bignum_negate(s9_cell a);
s9_cell	s9_bignum_shift_left(s9_cell a, int fill);
s9_cell	s9_bignum_shift_right(s9_cell a);
s9_cell	s9_bignum_subtract(s9_cell a, s9_cell b);
s9_cell	s9_bignum_to_int(s9_cell x, int *of);
s9_cell	s9_bignum_to_real(s9_cell a);
s9_cell	s9_bignum_to_string(s9_cell x);
int	s9_blockread(char *s, int k);
void	s9_blockwrite(char *s, int k);
void	s9_close_port(int port);
void	s9_close_input_string(void);
s9_cell	s9_cons3(s9_cell pcar, s9_cell pcdr, int ptag);
int	s9_conses(s9_cell a);
void	s9_cons_stats(int x);
s9_cell	s9_copy_string(s9_cell x);
void	s9_count(s9_counter *c);
void	s9_countn(s9_counter *c, int n);
char	*s9_dump_image(char *path, char *magic);
int	s9_error_port(void);
void	s9_exponent_chars(char *s);
void	s9_fatal(char *msg);
s9_cell	s9_find_symbol(char *s);
s9_cell	s9_flat_copy(s9_cell n, s9_cell *lastp);
void	s9_flush(void);
int	s9_gc(void);
int	s9_gcv(void);
void	s9_gc_verbosity(int n);
void	s9_get_counters(s9_counter **nc, s9_counter **cc, s9_counter **vc,
			s9_counter **gc);
void	s9_mem_error_handler(void (*h)(int src));
void	s9_image_vars(s9_cell **v);
int	s9_input_port(void);
int	s9_inport_open_p(void);
int	s9_integer_string_p(char *s);
s9_cell	s9_intern_symbol(s9_cell y);
s9_cell	s9_int_to_bignum(int v);
int	s9_io_status(void);
void	s9_io_reset(void);
int	s9_length(s9_cell n);
char	*s9_load_image(char *path, char *magic);
int	s9_lock_port(int port);
s9_cell	s9_make_char(int c);
s9_cell	s9_make_integer(s9_cell i);
s9_cell	s9_make_norm_real(int flags, s9_cell exp, s9_cell mant);
s9_cell	s9_make_port(int portno, s9_cell type);
s9_cell	s9_make_primitive(S9_PRIM *p);
s9_cell	S9_make_real(int flags, s9_cell exp, s9_cell mant);
s9_cell	s9_make_real(int sign, s9_cell exp, s9_cell mant);
s9_cell	s9_make_string(char *s, int k);
s9_cell	s9_make_symbol(char *s, int k);
s9_cell	s9_make_vector(int k);
s9_cell	s9_mkfix(int i);
int	s9_new_port(void);
s9_cell	s9_new_vec(s9_cell type, int size);
int	s9_open_input_port(char *path);
char	*s9_open_input_string(char *s);
int	s9_open_output_port(char *path, int append);
int	s9_output_port(void);
int	s9_outport_open_p(void);
int	s9_port_eof(int p);
void	s9_prints(char *s);
int	s9_printer_limit(void);
void	s9_print_bignum(s9_cell n);
void	s9_print_expanded_real(s9_cell n);
void	s9_print_real(s9_cell n);
void	s9_print_sci_real(s9_cell n);
int	s9_readc(void);
s9_cell	s9_read_counter(s9_counter *c);
s9_cell	s9_real_abs(s9_cell a);
s9_cell	s9_real_add(s9_cell a, s9_cell b);
int	s9_real_approx_p(s9_cell a, s9_cell b);
s9_cell	s9_real_ceil(s9_cell x);
s9_cell	s9_real_divide(s9_cell a, s9_cell b);
int	s9_real_equal_p(s9_cell a, s9_cell b);
s9_cell	s9_real_exponent(s9_cell x);
s9_cell	s9_real_floor(s9_cell x);
s9_cell	s9_real_integer_p(s9_cell x);
int	s9_real_less_p(s9_cell a, s9_cell b);
s9_cell	s9_real_mantissa(s9_cell x);
s9_cell	s9_real_multiply(s9_cell a, s9_cell b);
s9_cell	s9_real_negate(s9_cell a);
s9_cell	s9_real_negative_p(s9_cell a);
s9_cell	s9_real_positive_p(s9_cell a);
s9_cell	s9_real_power(s9_cell x, s9_cell y);
s9_cell	s9_real_round(s9_cell x);
s9_cell	s9_real_sqrt(s9_cell x);
s9_cell	s9_real_subtract(s9_cell a, s9_cell b);
s9_cell	s9_real_to_bignum(s9_cell r);
s9_cell	s9_real_to_string(s9_cell r, int mode);
s9_cell	s9_real_trunc(s9_cell x);
s9_cell	s9_real_zero_p(s9_cell a);
void	s9_rejectc(int c);
void	s9_reset(void);
void	s9_reset_counter(s9_counter *c);
void	s9_reset_std_ports(void);
void	s9_run_stats(int x);
void	s9_fini(void);
void	s9_init(s9_cell **extroots, s9_cell *stack, int *stkptr);
s9_cell	s9_set_input_port(s9_cell port);
void	s9_set_node_limit(int k);
s9_cell	s9_set_output_port(s9_cell port);
void	s9_set_printer_limit(int k);
void	s9_set_vector_limit(int k);
int	s9_string_numeric_p(char *s);
s9_cell	s9_string_to_bignum(char *s);
s9_cell	s9_string_to_number(char *s);
s9_cell	s9_string_to_real(char *s);
s9_cell	s9_string_to_symbol(s9_cell x);
s9_cell	s9_symbol_ref(char *s);
s9_cell	s9_symbol_table(void);
s9_cell	s9_symbol_to_string(s9_cell x);
char	*s9_typecheck(s9_cell f);
int	s9_unlock_port(int port);
s9_cell	s9_unsave(int k);
void	s9_writec(int c);
